src/dmd/dtemplate.d

```diff
@@ -2992,6 +2991,27 @@ void functionResolve(ref MatchAccumulator m, Dsymbol dstart, Loc loc, Scope* sc,
                 }
                 else
                     continue;   // MATCH.nomatch
+
+                // need to check here whether the constructor is the member of a struct
+                // declaration that defines a copy constructor. This is already checked
+                // in the semantic of CtorDeclaration, however, when matching functions,
+                // the template instance is not expanded.
+                auto ad = fd.isThis();
+                auto sd = ad.isStructDeclaration();
+                immutable dim = tf.parameterList.length;
+                if (sd && sd.hasCopyCtor && (dim == 1 || (dim > 1 && tf.parameterList[1].defaultArg)) )
+                {
+                    auto param = tf.parameterList[0];
+                    if (!(param.storageClass & STC.ref_) && param.type.mutableOf().unSharedOf() == sd.type.mutableOf().unSharedOf())
+                    {
+                        .error(fd.loc, "Cannot define both an rvalue constructor and a copy constructor for `struct %s`", sd.toChars);
+                        .errorSupplemental(ti.loc, "Template instance `%s` creates a rvalue constructor for `struct %s`",
+                                ti.toPrettyChars(), sd.toChars());
+
+                        continue;
+                    }
+                }
```

Comment:
Copying this code isn't a good solution. Maybe this bug indicates that the checks done in `dsymbolsem.d` are not in the right location to begin with. But at the very least move the logic into a common function.

Reply:
The check in dsymbolsem is definitely in the right location. The problem is that function resolution does not fully instantiate a templated function (i.e. it does not perform any semantic analysis) to check whether it is a match or not. It just creates a dummy template instance to check the constraints. Bypassing semantic analysis is the problem here, that's why we need to redo the check.